### ISSUE TITLE
demultiplexing

### ISSUE THREAD

**Issue by jcw** [jeelabs/folie#7]
Folie can treat received data as coming from multiple streams.
A simple approach, would be to use `\<n>` to switch to stream `<n>` (n = 1..5, perhaps).
Can only work if received data is plain text, never binary data containing one of these codes.

**Comment by jcw** [comment]
Consider using [SLIP](https://en.wikipedia.org/wiki/Serial_Line_Internet_Protocol) escape sequence, which supports full 8-bit binary pass-through.

**Comment by jcw** [comment]
There's some overlap with #6 (telnet escapes): from folie to µC needs out-of-band info to control the DTR, RTS, and parity settings for firmware uploaing, while from µC to folie needs out of band info to switch streams and indicate which thread the following data is coming from.

It's a bit odd to use telnet escapes one way and slip the other. No clear solution comes to mind.

**Comment by jcw** [comment]
Yes, I think SLIP is the way to go. It needs to escape $C0 and $DB, but apart from that it's 100% transparent - and best of all: when not used, the remote end doesn't have to be SLIP-aware.

**Comment by jcw** [comment]
The question is how to demultiplex, and what to do with those streams. I'm trying to keep Folie as state- and config-less as possible (so that it always behaves the same, in any project).

One idea would to publish text preceded by `<$DB><n>` to an MQTT topic (perhaps `/folie/stream<n>` on localhost as default).

But that's not really a good match: MQTT is not really character-stream-based, and this approach would not have any packet structure, it would simply continue to stream to that topic until a new `<$DB><n2>` marker is received, switching it to stream n2. This introduces state across host and remote - probably not a good idea (i.e. on remote: "where will my next output be going?").

I'd like to allow each task in a Forth-based multi-tasking app to send out whatever text it wants to the console with `emit`, `type`, etc - with a lower layer detecting the task, and inserting stream-switching escapes automatically as needed.

The `<$DB><n>` mechanism in itself would be enough for that, but then what? I don't want to implement a "stream management" user interface in Folie, where you say: stream 1 goes to the console, stream 2 goes to filex, stream 3 gets sent to a net socket, stream 4 goes to a named pipe, etc.

**Comment by jcw** [comment]
Hey, @tve and @tht, maybe you have some thoughts about the above https://github.com/jeelabs/folie/issues/7#issuecomment-274183474 ?

**Comment by JohnOH** [comment]
Does SLIP not allow the endpoint to use IP socket numbers  and then folie to pass the socket stream to the routed by the OS?

**Comment by jcw** [comment]
Ah, no , I'm not transferring TCP/IP packets here - just using SLIP as serial line encoding mechanism, to be able to delimit packet boundaries and send some other messages. Doing TCP/IP would require implementing an TCP/IP stack on the remote end.

**Comment by tve** [comment]
I'm not sure what the use-cases here are. Lemme try:

1. multiple tasks printing debug info, would be nice to pretty-print it, i.e., ensure it doesn't get completely garbled
2. input directed at the REPL
3. input directed at any task
4. tasks outputting MQTT messages
5. incoming MQTT messages forwarded to tasks
6. other stream coming out of a task being forwarded on some unix file descriptor (file, pipe, socket, ...)
7. other stream coming in from some unix FD being forwarded to a task
8. ???

Which of these do you want to solve in folie? ;-)

Here's a stab:
- each SLIP pkt has a stream ID, chars outside of a packet are assigned stream ID 0
- each stream has a type that gets set via explicit command in the stream
- task "std" output gets assigned a line-buffered type
- repl output gets assigned either a line-buffered or special repl type
- folie uses type to pretty-print or route output
- if a task wants to output mqtt, it has to change the type of its output or create a new stream and set its type
- stdin goes to the repl stream
- the mqtt stream type supports commands, such as subscribe
- add as much additional complexity as you feel compelled to... ;-)

**Comment by jcw** [comment]
Thx. Now I need to think about whether tasks are the best way forward - they are relatively expensive in terms of RAM use. Another option would be to have two or three fixed tasks, not per-app: one for the REPL, one for background activity, timer management, maybe even GC, who knows. Possibly a third task for application-specific (foreground?) use.

**Comment by JohnOH** [comment]
Enumerating multiple comm ports seems like a  reasonably hardware independent approach, prior art in the BMP code.

**Comment by jcw** [comment]
> Which of these do you want to solve in folie? ;-)

My main goal is to keep a responsive command interpreter as much as possible, even when "the app" is doing stuff in the background. Very much like the JS console in the browser or in Node (or Tcl, for that matter). That's essentially just another way of saying that things need to be event-driven.

At 500+ bytes of RAM per task (with 64-word data and return stacks), things could get a bit cramped on an STML052 if everything ends up being a separate task. I think I'd much rather have JavaScript's `setTimeout` and `setInterval` callback mechanism, that should be sufficient for most repetitive remote node activities, as well as to drive a PureData-like engine.

I think I'll go for two tasks: a foreground task being the Mecrisp command interpreter, and a background task for everything else. With `multi.fs` it's easy to implement those, and to enable/disable the background activity while fooling around with the code. It's also very easy to see which of the two tasks is running (`: fg? up @ boot-task = ;`). That in turn, allows implementing a version of `emit` which injects a "task prefix byte" into the output stream when the output is not coming from the same task as before.

So my plan now is to insert CTRL-F and CTRL-B bytes into the output stream, and to de-multiplex this in Folie. This does assume that binary data is never emitted. Not sure yet if that's good enough - it could be solved by using SLIP or Telnet escapes as markers instead.

A simple modified `hook-emit` word would be all that's needed to implement this: it compares `fg?` to the last known state and injects a prefix whenever it has changed.

The result: Folie always sends text assuming that it will be consumed by the Mecrisp (foreground) command interpreter, and splits all incoming data (i.e. outbound from Mecrisp) into two streams - one is treated as responses from Mecrisp, i.e. command echo and replies to command requests, the other is for asynchronous "events", and could be sent into MQTT (one event per line) for example.

This is about as simple as I can make it to get an async Node-like command prompt, I think.

**Comment by jcw** [comment]
FWIW, none of the above strikes me as particularly elegant. _Yet another ad-hoc mechanism..._

**Comment by JohnOH** [comment]
Are we certain that enumerating a second USB logical port would not be more *open* other user level code. Are we fixed to using serial, might not a more binary type interface be interesting.

**Comment by jcw** [comment]
Yes, you have a point. I was going to say that this won't solve the issue for when a serial rf-to-rf console gets implemented, but I now realise that the node could easily be made to talk in two different ways (header bits, or net group) to separate app comms from the interactive prompt.

I'm not looking forward to the prospect of adding stuff to the `suf/usb.fs` code, but that's just me. If there's a way to do it (like the Black Magic Probe's two serial ports), it could be worthwhile after all.
